### PR TITLE
Sharing: Do not show copy action when user doesn't have permissions

### LINKED ISSUE #37729: [Bug]: File copy action is visible even though user doesn't have permission
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

File copy action is displayed even if the user doesn't have necessary permissions. Trying to copy the shared file is still getting rejected properly.

### Steps to reproduce

1. Open using another account
2. Share a folder to test account
3. Uncheck all rights
4. Go to web app via test account
5. Go to the shared folder
6. Click on the button "..." next to a file inside the shared folder
7. Obtained Results : Also have "Copy" action

### Expected behavior

Expected results : Should not see Copy action

### Installation method

Community Docker image

### Nextcloud Server version

master

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
Enabled:
  - cloud_federation_api: 1.10.0
  - comments: 1.17.0
  - contactsinteraction: 1.8.0
  - dashboard: 7.7.0
  - dav: 1.26.0
  - federatedfilesharing: 1.17.0
  - federation: 1.17.0
  - files: 1.22.0
  - files_sharing: 1.19.0
  - files_trashbin: 1.17.0
  - files_versions: 1.20.0
  - lookup_server_connector: 1.15.0
  - oauth2: 1.15.0
  - photos: 2.0.1
  - profiler: 1.3.0
  - provisioning_api: 1.17.0
  - settings: 1.9.0
  - sharebymail: 1.17.0
  - systemtags: 1.17.0
  - text: 3.7.2
  - theming: 2.2.0
  - twofactor_backupcodes: 1.16.0
  - updatenotification: 1.17.0
  - user_status: 1.7.0
  - viewer: 2.1.0
  - weather_status: 1.7.0
  - workflowengine: 2.9.0
Disabled:
  - admin_audit: 1.17.0
  - bookmarks: 13.0.0
  - deck: 1.8.1
  - encryption: 2.15.0
  - files_external: 1.19.0
  - recognize: 3.7.0
  - richdocuments: 8.1.0-dev.1
  - tables: 0.3.0
  - testing: 1.17.0
  - user_ldap: 1.17.0
  - user_oidc: 1.3.1
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by marcelklehr**
fixes #37729

* Resolves: #37729

<details>
<summary>Screencasts</summary>

Before
Peek 2023-04-18 18-20

After
Peek 2023-04-18 18-18

</details>

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by artonge — CHANGES_REQUESTED** [review]
Shouldn't the user be allowed to copy a file if he can read it?

**Comment by artonge** [comment]
No, it is rather technical. If I can read the file, then copying is possible, either by copy-pasting, or if the software allows it, with one click.

I also just check and Nautilus is allowing copying read-only files. Same with bash.

**Comment by skjnldsv** [comment]
> or we fix it in the backend to make it possible to copy files when the user has read permissions. I'll let you fight this out with @jancborchardt

I think this is the way to go as this is what I would expect from a dav endpoint.
The question here is where does it fail? When you copy, does it fail on the READ check, or on the WRITE permission of the destination folder?

**Comment by marcelklehr** [comment]
Seems like the permission error stems from here: https://github.com/nextcloud/server/blob/master/apps/dav/lib/DAV/ViewOnlyPlugin.php

**Comment by jancborchardt** [comment]
Sorry, I assumed the backend was as intended. If not, then I would agree that this is rather something to fix in the backend. :)

File: apps/files_sharing/src/share.js [hunk 1]
```diff
@@ -94,7 +94,7 @@ import { getCapabilities } from '@nextcloud/capabilities'
 				}
 				if (_.isFunction(fileData.canDownload) && !fileData.canDownload()) {
 					delete fileActions.actions.all.Download
-					if (fileData.permissions & OC.PERMISSION_UPDATE === 0) {
+					if ((fileData.permissions & OC.PERMISSION_UPDATE) === 0) {
```
Comment (marcelklehr) [inline apps/files_sharing/src/share.js:97]:
```
> (1 & 2 === 0)
0
> (1 & 2) === 0
true
```

:rage:

**Comment by marcelklehr** [comment]
It turns out, there was already code for this, but it didn't work.

File: apps/files_sharing/src/share.js (hunk 1 above)
Comment (come-nc) [reply apps/files_sharing/src/share.js:97]: I always feel we do not put enough parenthesis in Nextcloud code, I tend to be paranoid about these things. There are so many operators it’s too hard to remember all priorities by heart.

**Review by artonge — APPROVED** [review]

**Review by come-nc — APPROVED** [review]

**Comment by solracsf** [comment]
/backport https://github.com/nextcloud/server/commit/af665378f8428963f899f568a6cd5de9e233e82e to stable27

**Pull request merged**